File: back/src/main/java/com/triptrace/domain/image/image/service/ImageService.java
package com.triptrace.domain.image.image.service;

import java.util.List;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import com.triptrace.domain.image.image.error.ImageErrorCode;
import com.triptrace.domain.image.image.dto.response.ImageServiceResponse;
import com.triptrace.domain.image.image.entity.Image;
import com.triptrace.domain.image.image.mapper.ImageMapper;
import com.triptrace.domain.image.image.repository.ImageRepository;
import com.triptrace.domain.marker.marker.repository.MarkerRepository;
import com.triptrace.domain.member.member.entity.Member;
import com.triptrace.domain.post.post.entity.Post;
import com.triptrace.domain.trip.trip.entity.Trip;
import com.triptrace.domain.trip.trip.repository.TripRepository;
import com.triptrace.global.exception.ServiceException;

@Service
public class ImageService {
    private final ImageRepository imageRepository;
    private final TripRepository tripRepository;
    private final MarkerRepository markerRepository;

    @Transactional
    public ImageServiceResponse create(Image image) {
        image = imageRepository.save(image);
        return ImageMapper.toServiceResponse(image);
    }

    @Transactional
    public ImageServiceResponse delete(Image image) {
        ImageServiceResponse response = ImageMapper.toServiceResponse(image);
        imageRepository.delete(image);
        return response;
    }

    @Transactional
    public ImageServiceResponse modifyPost(Member owner, Trip trip, Post post, Long imageId) {
        Image image = getById(imageId);
        if (!validateOwner(owner, image)) {
            throw new ServiceException(ImageErrorCode.FORBIDDEN);
        }
        if (!validateTrip(trip, image)) {
            throw new ServiceException(ImageErrorCode.INVALID_TRIP);
        }
        image.modifyPost(post);
        return ImageMapper.toServiceResponse(image);
    }

    @Transactional
    public ImageServiceResponse delete(Member owner, Trip trip, Post post, Long id) {
        Image image = getById(id);
        validate(owner, trip, post, image);
        disconnectRepresentativeReferences(image.getId());
        return delete(image);
    }

    @Transactional
    public ImageServiceResponse delete(Member owner, Trip trip, Long id) {
        Image image = getById(id);
        validate(owner, trip, null, image);
        disconnectRepresentativeReferences(image.getId());
        return delete(image);
    }

    @Transactional
    public ImageServiceResponse delete(Member owner, Trip trip, Post post, String imageUrl) {
        Image image = getByUrl(imageUrl);
        validate(owner, trip, post, image);
        disconnectRepresentativeReferences(image.getId());
        return delete(image);
    }

    private void disconnectRepresentativeReferences(Long imageId) {
        tripRepository.findByRepresentativeImageId(imageId)
            .forEach(trip -> trip.changeRepresentativeImage(null));
        markerRepository.findByRepresentativeImageId(imageId)
            .forEach(marker -> marker.changeRepresentativeImage(null));
    }

    @Transactional(readOnly = true)
    public Image getById(Long id) {
        return imageRepository.findById(id)
            .orElseThrow(() -> new ServiceException(ImageErrorCode.NOT_FOUND));
    }

    @Transactional(readOnly = true)
    public Image getByUrl(String originalFileUrl) {
        return imageRepository.findByOriginalFileUrl(originalFileUrl)
            .orElseThrow(() -> new ServiceException(ImageErrorCode.NOT_FOUND));
    }

    @Transactional(readOnly = true)
    public ImageServiceResponse findById(Long id) {
        Image image = getById(id);
        return ImageMapper.toServiceResponse(image);
    }

    @Transactional(readOnly = true)
    public ImageServiceResponse findByUrl(String imageUrl) {
        Image image = getByUrl(imageUrl);
        return ImageMapper.toServiceResponse(image);
    }

    @Transactional(readOnly = true)
    public List<ImageServiceResponse> findWithOwner(Long ownerId) {
        List<Image> images = imageRepository.findByOwnerId(ownerId);
        return images.stream().map(ImageMapper::toServiceResponse).toList();
    }

    @Transactional(readOnly = true)
    public List<ImageServiceResponse> findByTripId(Trip trip) {
        List<Image> images = imageRepository.findByTripId(trip.getId());
        return images.stream().map(ImageMapper::toServiceResponse).toList();
    }

    @Transactional(readOnly = true)
    public List<ImageServiceResponse> findByPostId(Post post) {
        List<Image> images = imageRepository.findByPostId(post.getId());
        return images.stream().map(ImageMapper::toServiceResponse).toList();
    }

    private void validate(Member owner, Trip trip, Post post, Image image) {
        if (!validateOwner(owner, image)) {
            throw new ServiceException(ImageErrorCode.FORBIDDEN);
        }
        if (!validateTrip(trip, image)) {
            throw new ServiceException(ImageErrorCode.INVALID_TRIP);
        }
        if (!validatePost(post, image)) {
            throw new ServiceException(ImageErrorCode.INVALID_POST);
        }
    }

    private boolean validateOwner(Member owner, Image image) {
        return owner.getId().equals(image.getOwner().getId());
    }

    private boolean validateTrip(Trip trip, Image image) {
        return trip.getId().equals(image.getTrip().getId());
    }

    private boolean validatePost(Post post, Image image) {
        if (post == null) {
            return true;
        }
        if (image.getPost() == null) {
            return false;
        }
        return post.getId().equals(image.getPost().getId());
    }

    @Transactional
    public ImageServiceResponse unassign(Long ownerId, Long tripId, Long imageId) {
        Image image = imageRepository
            .findByIdAndOwnerIdAndTripId(imageId, ownerId, tripId)
            .orElseThrow(() -> new ServiceException(ImageErrorCode.NOT_FOUND));
        image.modifyPost(null);
        return ImageMapper.toServiceResponse(image);
    }

    @java.lang.SuppressWarnings("all")
    public ImageService(final ImageRepository imageRepository, final TripRepository tripRepository, final MarkerRepository markerRepository) {
        this.imageRepository = imageRepository;
        this.tripRepository = tripRepository;
        this.markerRepository = markerRepository;
    }
}

File: back/src/main/java/com/triptrace/domain/image/image/service/ImageService.kt
package com.triptrace.domain.image.image.service

import com.triptrace.domain.image.image.dto.response.ImageServiceResponse
import com.triptrace.domain.image.image.entity.Image
import com.triptrace.domain.image.image.error.ImageErrorCode
import com.triptrace.domain.image.image.mapper.ImageMapper
import com.triptrace.domain.image.image.repository.ImageRepository
import com.triptrace.domain.marker.marker.repository.MarkerRepository
import com.triptrace.domain.member.member.entity.Member
import com.triptrace.domain.post.post.entity.Post
import com.triptrace.domain.trip.trip.entity.Trip
import com.triptrace.domain.trip.trip.repository.TripRepository
import com.triptrace.global.exception.ServiceException
import org.springframework.stereotype.Service
import org.springframework.data.repository.findByIdOrNull
import org.springframework.transaction.annotation.Transactional

@Service
class ImageService(
    private val imageRepository: ImageRepository,
    private val tripRepository: TripRepository,
    private val markerRepository: MarkerRepository,
) {
    @Transactional
    fun create(image: Image): ImageServiceResponse =
        ImageMapper.toServiceResponse(imageRepository.save(image))

    @Transactional
    fun delete(image: Image): ImageServiceResponse {
        val response = ImageMapper.toServiceResponse(image)
        imageRepository.delete(image)
        return response
    }

    @Transactional
    fun modifyPost(owner: Member, trip: Trip, post: Post, imageId: Long): ImageServiceResponse {
        val image = getById(imageId)
        if (!validateOwner(owner, image)) throw ServiceException(ImageErrorCode.FORBIDDEN)
        if (!validateTrip(trip, image)) throw ServiceException(ImageErrorCode.INVALID_TRIP)
        image.modifyPost(post)
        return ImageMapper.toServiceResponse(image)
    }

    @Transactional
    fun delete(owner: Member, trip: Trip, post: Post?, id: Long): ImageServiceResponse {
        val image = getById(id)
        validate(owner, trip, post, image)
        disconnectRepresentativeReferences(image.id)
        return delete(image)
    }

    @Transactional
    fun delete(owner: Member, trip: Trip, id: Long): ImageServiceResponse {
        val image = getById(id)
        validate(owner, trip, null, image)
        disconnectRepresentativeReferences(image.id)
        return delete(image)
    }

    @Transactional
    fun delete(owner: Member, trip: Trip, post: Post?, imageUrl: String): ImageServiceResponse {
        val image = getByUrl(imageUrl)
        validate(owner, trip, post, image)
        disconnectRepresentativeReferences(image.id)
        return delete(image)
    }

    @Transactional(readOnly = true)
    fun getById(id: Long): Image = imageRepository.findByIdOrNull(id)
        ?: throw ServiceException(ImageErrorCode.NOT_FOUND)

    @Transactional(readOnly = true)
    fun getByUrl(originalFileUrl: String): Image =
        imageRepository.findByOriginalFileUrl(originalFileUrl)
            ?: throw ServiceException(ImageErrorCode.NOT_FOUND)

    @Transactional(readOnly = true)
    fun findById(id: Long): ImageServiceResponse = ImageMapper.toServiceResponse(getById(id))

    @Transactional(readOnly = true)
    fun findWithOwner(ownerId: Long): List<ImageServiceResponse> =
        imageRepository.findByOwnerId(ownerId)
            .map(ImageMapper::toServiceResponse)

    @Transactional
    fun unassign(ownerId: Long, tripId: Long, imageId: Long): ImageServiceResponse {
        val image = imageRepository.findByIdAndOwnerIdAndTripId(imageId, ownerId, tripId)
            ?: throw ServiceException(ImageErrorCode.NOT_FOUND)
        image.modifyPost(null)
        return ImageMapper.toServiceResponse(image)
    }

    private fun disconnectRepresentativeReferences(imageId: Long?) {
        if (imageId == null) return
        tripRepository.findByRepresentativeImageId(imageId)
            .forEach { it.changeRepresentativeImage(null) }
        markerRepository.findByRepresentativeImageId(imageId)
            .forEach { it.changeRepresentativeImage(null) }
    }

    private fun validate(owner: Member, trip: Trip, post: Post?, image: Image) {
        if (!validateOwner(owner, image)) throw ServiceException(ImageErrorCode.FORBIDDEN)
        if (!validateTrip(trip, image)) throw ServiceException(ImageErrorCode.INVALID_TRIP)
        if (!validatePost(post, image)) throw ServiceException(ImageErrorCode.INVALID_POST)
    }

    private fun validateOwner(owner: Member, image: Image): Boolean = owner.id == image.owner.id

    private fun validateTrip(trip: Trip, image: Image): Boolean = trip.id == image.trip.id

    private fun validatePost(post: Post?, image: Image): Boolean =
        post == null || (image.post != null && post.id == image.post?.id)
}
